src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,15 +300,44 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app max-command-failure? max-command-size]
+  (fn [req]
+    (if (= :post (:request-method req))
+      (let [length-in-bytes (request/content-length req)]
+
+        (if length-in-bytes
+          (log/infof "Processing command with a content-length of %s bytes" length-in-bytes )
+          (log/warn (str "No content length found for POST. "
+                         "POST bodies that are too large could cause memory-related failures.")))
+
+        (if (and length-in-bytes
+                 max-command-failure?
+                 (> length-in-bytes max-command-size))
+          (do
+            (log/warnf "content-length of command is %sMB and is larger than the maximum allowed command size of %sMB"
+                       (utils/bytes->mb length-in-bytes)
+                       (utils/bytes->mb max-command-size))
+            (.close (:body req))
```

Comment:
So from the puppet master (running using webrick to be clear) we get the error message below, its basically 'broken pipe', I don't think its seeing a 413, definitely not presenting the error message we've provided later on anyway.

```
Info: 'replace facts' command for macbook-pro.local submitted to PuppetDB with UUID 71bd7592-31d3-4963-ad2c-2ae109e6f476
Debug: Using cached facts for macbook-pro.local
Info: Caching node for macbook-pro.local
Notice: Compiled catalog for macbook-pro.local in environment production in 30.09 seconds
Info: Caching catalog for macbook-pro.local
Debug: Creating new connection for https://puppet:8081
Warning: Error connecting to puppet on 8081 at route /pdb/cmd/v1?checksum=95f7fbca7440767b134a1ddb748db27552e0152d, error message received was 'Broken pipe'. Failing over to the next PuppetDB url in the 'server_urls' list
Error: Failed to execute '/pdb/cmd/v1?checksum=95f7fbca7440767b134a1ddb748db27552e0152d' on any of the following 'server_urls': https://puppet:8081
Error: /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/http.rb:102:in `action' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:61:in `block in submit' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:60:in `submit' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:67:in `block in submit_command' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:64:in `submit_command' puppet/lib/puppet/indirector/catalog/puppetdb.rb:13:in `block in save' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile' /Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile' puppet/lib/puppet/indirector/catalog/puppetdb.rb:11:in `save' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/store_configs.rb:24:in `save' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/indirection.rb:200:in `find' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:134:in `do_find' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:48:in `block in call' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/context.rb:65:in `override' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet.rb:240:in `override' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:47:in `call' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:82:in `block in process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `each' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:60:in `block in process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:58:in `process' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:49:in `block in service' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `synchronize' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `service' /Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:138:in `service' /Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:94:in `run' /Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick.rb:33:in `block (2 levels) in listen' /Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `call' /Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `block in start_thread'
Error: Failed to execute '/pdb/cmd/v1?checksum=95f7fbca7440767b134a1ddb748db27552e0152d' on any of the following 'server_urls': https://puppet:8081
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:94:in `rescue in submit'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:59:in `submit'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:67:in `block in submit_command'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:64:in `submit_command'
/Users/ken/Development/puppetdb/puppet/lib/puppet/indirector/catalog/puppetdb.rb:13:in `block in save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/indirector/catalog/puppetdb.rb:11:in `save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/store_configs.rb:24:in `save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/indirection.rb:200:in `find'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:134:in `do_find'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:48:in `block in call'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/context.rb:65:in `override'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet.rb:240:in `override'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:47:in `call'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:82:in `block in process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `each'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:60:in `block in process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:58:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:49:in `block in service'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `synchronize'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `service'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:138:in `service'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:94:in `run'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick.rb:33:in `block (2 levels) in listen'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `call'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `block in start_thread'
Error: Failed to execute '/pdb/cmd/v1?checksum=95f7fbca7440767b134a1ddb748db27552e0152d' on any of the following 'server_urls': https://puppet:8081
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:94:in `rescue in submit'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb/command.rb:59:in `submit'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:67:in `block in submit_command'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:64:in `submit_command'
/Users/ken/Development/puppetdb/puppet/lib/puppet/indirector/catalog/puppetdb.rb:13:in `block in save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/util/puppetdb.rb:89:in `profile'
/Users/ken/Development/puppetdb/puppet/lib/puppet/indirector/catalog/puppetdb.rb:11:in `save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/store_configs.rb:24:in `save'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/indirector/indirection.rb:200:in `find'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:134:in `do_find'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:48:in `block in call'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/context.rb:65:in `override'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet.rb:240:in `override'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/api/indirected_routes.rb:47:in `call'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:82:in `block in process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `each'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:81:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/route.rb:87:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:60:in `block in process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler/around_profiler.rb:58:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/util/profiler.rb:51:in `profile'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/handler.rb:58:in `process'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:49:in `block in service'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `synchronize'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick/rest.rb:48:in `service'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:138:in `service'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/httpserver.rb:94:in `run'
/Users/ken/.rvm/gems/ruby-2.0.0-p647@puppetdb/gems/puppet-4.3.1/lib/puppet/network/http/webrick.rb:33:in `block (2 levels) in listen'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `call'
/Users/ken/.rvm/rubies/ruby-2.0.0-p647/lib/ruby/2.0.0/webrick/server.rb:295:in `block in start_thread'
Debug: Routes Registered:
Debug: Route /^\/puppet\//
Debug: Route /^\/puppet-ca\//
Debug: Route /.*/
```